hypervisor/boot/guest/vboot_info.c

```diff
@@ -28,66 +28,68 @@
  * but this should be configurable for different OS. */
 #define DEFAULT_RAMDISK_GPA_MAX		0x37ffffffUL
 
+#define PRE_VM_MAX_RAM_ADDR_BELOW_4GB		(VIRT_ACPI_DATA_ADDR - 1U)
+
 /**
  * @pre vm != NULL && mod != NULL
  */
 static void init_vm_ramdisk_info(struct acrn_vm *vm, const struct abi_module *mod)
 {
 	uint64_t ramdisk_load_gpa = INVALID_GPA;
+	uint64_t ramdisk_gpa_max = DEFAULT_RAMDISK_GPA_MAX;
+	uint64_t kernel_start = (uint64_t)vm->sw.kernel_info.kernel_load_addr;
+	uint64_t kernel_end = kernel_start + vm->sw.kernel_info.kernel_size;
 	struct acrn_vm_config *vm_config = get_vm_config(vm->vm_id);
 
 	if (mod->start != NULL) {
 		vm->sw.ramdisk_info.src_addr = mod->start;
 		vm->sw.ramdisk_info.size = mod->size;
 	}
 
-	if (is_sos_vm(vm)) {
+	/* Per Linux boot protocol, the Kernel need a size of contiguous
+	 * memory(i.e. init_size field in zeropage) from its extract address to boot,
+	 * and initrd_addr_max field specifies the maximum address of the ramdisk.
+	 * Per kernel src head_64.S, decompressed kernel start at 2M aligned to the
+	 * compressed kernel load address.
+	 */
+	if (vm->sw.kernel_type == KERNEL_BZIMAGE) {
 		struct zero_page *zeropage = (struct zero_page *)vm->sw.kernel_info.kernel_src_addr;
+		uint32_t kernel_init_size = zeropage->hdr.init_size;
 		uint32_t initrd_addr_max = zeropage->hdr.initrd_addr_max;
 
+		kernel_end = kernel_start + MEM_2M + kernel_init_size;
```

Comment:
here you use different algorithm compare to get_kernel_load_addr？

Reply:
in get_kernel_load_addr, we need to ensure kernel load addr to be kernel_aligned, so we enlarge the space to (kernel_init_size + 2 * kernel_align), but here kernel load address is sure to be kernel_aligned, so we can get more precise kernel end address.